inline postgres certificates unsupported

I deploy a web app to a local cloudfoundry environment. As a database service for my DEV environment I have chosen a Marketplace service google-cloudsql-postgres with the plan postgres-db-f1-micro. Using the Web UI I created an instance with the name myapp-test-database and mentioned it in the CF Manifest:

    applications:
    - name: myapp-test
      services:
       - myapp-test-database

With this configuration the Mendix Runtime fails with:

    ERROR - ConnectionBus: Opening JDBC connection to Some(34.91.242.101) failed with SQLState: 08006 Error code: 0 Message: Could not open SSL certificate file -----BEGIN CERTIFICATE-----
    ...
    Caused by: java.io.FileNotFoundException: -----BEGIN CERTIFICATE-----
    ...
    -----END CERTIFICATE----- (File name too long)

It's because the JDBC driver does not support inline certificates in `DATABASE_URL=postgres://user:pass@host?sslcert=-----BEGIN+CERTIFICATE-inlineCertificate... `

Here: https://github.com/mendix/cf-mendix-buildpack/blob/ce60432790e501e4237cbeca62b1a33c4849a77c/buildpack/runtime_components/database.py#L347
the jdbc params need to be parsed further, the inline certificates extracted to disk and PKCS#1 PEM key converted to PKCS#8 binary key that Postgres driver understands.

--- a/Makefile
+++ b/Makefile
@@ -13,7 +13,7 @@ download_wheels: requirements create_build_dirs
 	rm -rf build/vendor/wheels
 	mkdir -p build/vendor/wheels
 	pip3 download -d build/vendor/wheels/ --only-binary :all: pip setuptools wheel
-	pip3 download -d build/vendor/wheels/ --no-deps --platform manylinux1_x86_64 --python-version 36 -r requirements.txt
+	pip3 download -d build/vendor/wheels/ --no-deps --platform manylinux2010_x86_64 --python-version 36 -r requirements.txt
 
 .PHONY: create_build_dirs
 create_build_dirs:

--- a/buildpack/runtime_components/database.py
+++ b/buildpack/runtime_components/database.py
@@ -3,6 +3,8 @@ import logging
 import os
 import re
 import shutil
+import subprocess
+import tempfile
 from abc import ABC, abstractmethod
 from urllib.parse import parse_qs, unquote, urlencode
 
@@ -247,6 +249,12 @@ class DatabaseConfiguration(ABC):
 class UrlDatabaseConfiguration(DatabaseConfiguration):
     # Returns a database configuration based on the original code from util.
 
+    SSLROOTCERT = "sslrootcert"
+
+    SSLKEY = "sslkey"
+
+    SSLCERT = "sslcert"
+
     def __init__(self, url, env_vars=None):
         super().__init__(env_vars)
         logging.debug("Detected URL based database configuration.")
@@ -319,6 +327,11 @@ class UrlDatabaseConfiguration(DatabaseConfiguration):
             )
             jdbc_params.update({"sslmode": "verify-full"})
 
+        if database_type == "PostgreSQL":
+            self.extract_inline_cert(jdbc_params, self.SSLCERT, "postgresql.crt");
+            self.extract_inline_cert(jdbc_params, self.SSLKEY, "postgresql.pk8");
+            self.extract_inline_cert(jdbc_params, self.SSLROOTCERT, "root.crt");
+
         extra_url_params_str = self.env_vars.get(
             "DATABASE_CONNECTION_PARAMS", "{}"
         )
@@ -343,6 +356,39 @@ class UrlDatabaseConfiguration(DatabaseConfiguration):
                 config.update({"DatabaseUseSsl": True})
 
         self.m2ee_config = config
+
+    def extract_inline_cert(self, jdbc_params, param, filename):
+        list = jdbc_params.get(param)
+        if list is not None:
+            for i,inline in enumerate(list):
+                if inline.startswith("-----BEGIN"):
+                    tmp_cert_file = self.create_cert_file(filename)
+                    try:
+                        logging.debug("extract %s to %s", param, tmp_cert_file.name)
+                        if inline.startswith("-----BEGIN RSA PRIVATE KEY-----"):
+                            self.to_pk8(inline, tmp_cert_file)
+                        else:
+                            tmp_cert_file.write(inline)
+                    finally:
+                        tmp_cert_file.close()
+                    list[i] = self.to_java_path(tmp_cert_file.name)
+
+    def to_java_path(self, file_path):
+        # Convert absolute path to the java subsystem
+        return file_path
+
+    def to_pk8(self, inline, tmp_cert_file):
+        tmp_cert_file.close()
+        dst_path = tmp_cert_file.name
+        # Cannot pass absolute path to openssl possibly belonging to a different subsystem
+        (dst_dir, dst_name) = os.path.split(dst_path)
+        subprocess.run(
+            ["openssl", "pkcs8", "-topk8", "-outform", "DER", "-inform", "PEM", "-out", dst_name, "-nocrypt"],
+            input=inline, cwd=dst_dir, encoding="ascii").check_returncode()
+
+    def create_cert_file(self, default_file_name):
+        (prefix, suffix) = os.path.splitext(default_file_name)
+        return tempfile.NamedTemporaryFile(mode="w",suffix=suffix,prefix=prefix,delete=False)
 
     def get_jdbc_strings(self, url, config, jdbc_params):
         # JDBC strings might be different from connection uri strings

--- a/buildpack/runtime_components/database.py
+++ b/buildpack/runtime_components/database.py
@@ -3,9 +3,9 @@ import logging
 import os
 import re
 import shutil
-import subprocess
 import tempfile
 from abc import ABC, abstractmethod
+from cryptography.hazmat.primitives import serialization
 from urllib.parse import parse_qs, unquote, urlencode
 
 from buildpack import util
@@ -327,10 +327,14 @@ class UrlDatabaseConfiguration(DatabaseConfiguration):
             )
             jdbc_params.update({"sslmode": "verify-full"})
 
-        if database_type == "PostgreSQL":
-            self.extract_inline_cert(jdbc_params, self.SSLCERT, "postgresql.crt");
-            self.extract_inline_cert(jdbc_params, self.SSLKEY, "postgresql.pk8");
-            self.extract_inline_cert(jdbc_params, self.SSLROOTCERT, "root.crt");
+        if database_type == "PostgreSQL" and not self.url.startswith("jdbc:"):
+            self.extract_inline_cert(
+                jdbc_params, self.SSLCERT, "postgresql.crt"
+            )
+            self.extract_inline_cert(
+                jdbc_params, self.SSLKEY, "postgresql.pk8"
+            )
+            self.extract_inline_cert(jdbc_params, self.SSLROOTCERT, "root.crt")
 
         extra_url_params_str = self.env_vars.get(
             "DATABASE_CONNECTION_PARAMS", "{}"
@@ -358,37 +362,51 @@ class UrlDatabaseConfiguration(DatabaseConfiguration):
         self.m2ee_config = config
 
     def extract_inline_cert(self, jdbc_params, param, filename):
-        list = jdbc_params.get(param)
-        if list is not None:
-            for i,inline in enumerate(list):
+        lst = jdbc_params.get(param)
+        if lst is not None:
+            if not isinstance(lst, list):
+                lst = [lst]
+                jdbc_params[param] = lst
+            for i, inline in enumerate(lst):
                 if inline.startswith("-----BEGIN"):
                     tmp_cert_file = self.create_cert_file(filename)
                     try:
-                        logging.debug("extract %s to %s", param, tmp_cert_file.name)
-                        if inline.startswith("-----BEGIN RSA PRIVATE KEY-----"):
+                        logging.debug(
+                            "extract %s to %s", param, tmp_cert_file.name
+                        )
+                        if inline.startswith(
+                            "-----BEGIN RSA PRIVATE KEY-----"
+                        ):
                             self.to_pk8(inline, tmp_cert_file)
                         else:
-                            tmp_cert_file.write(inline)
+                            tmp_cert_file.write(inline.encode())
                     finally:
                         tmp_cert_file.close()
-                    list[i] = self.to_java_path(tmp_cert_file.name)
+                    lst[i] = self.to_java_path(tmp_cert_file.name)
 
-    def to_java_path(self, file_path):
+    @classmethod
+    def to_java_path(cls, file_path):
         # Convert absolute path to the java subsystem
         return file_path
 
-    def to_pk8(self, inline, tmp_cert_file):
-        tmp_cert_file.close()
-        dst_path = tmp_cert_file.name
-        # Cannot pass absolute path to openssl possibly belonging to a different subsystem
-        (dst_dir, dst_name) = os.path.split(dst_path)
-        subprocess.run(
-            ["openssl", "pkcs8", "-topk8", "-outform", "DER", "-inform", "PEM", "-out", dst_name, "-nocrypt"],
-            input=inline, cwd=dst_dir, encoding="ascii").check_returncode()
+    @classmethod
+    def to_pk8(cls, inline, tmp_cert_file):
+        private_key = serialization.load_pem_private_key(
+            inline.encode(), password=None
+        )
+        keybytes = private_key.private_bytes(
+            encoding=serialization.Encoding.DER,
+            format=serialization.PrivateFormat.PKCS8,
+            encryption_algorithm=serialization.NoEncryption(),
+        )
+        tmp_cert_file.write(keybytes)
 
-    def create_cert_file(self, default_file_name):
+    @classmethod
+    def create_cert_file(cls, default_file_name):
         (prefix, suffix) = os.path.splitext(default_file_name)
-        return tempfile.NamedTemporaryFile(mode="w",suffix=suffix,prefix=prefix,delete=False)
+        return tempfile.NamedTemporaryFile(
+            mode="wb", suffix=suffix, prefix=prefix, delete=False
+        )
 
     def get_jdbc_strings(self, url, config, jdbc_params):
         # JDBC strings might be different from connection uri strings

--- a/requirements-dev.in
+++ b/requirements-dev.in
@@ -1,5 +1,4 @@
 black==19.10b0
-cryptography==3.2.1
 idna==2.8
 nose-timer==1.0.0
 pylint==2.5.3

--- a/requirements.in
+++ b/requirements.in
@@ -1,5 +1,6 @@
 backoff==1.10.0
 certifi==2020.12.5
+cryptography==3.4.7
 httplib2==0.19.0
 jinja2==2.11.3
 omegaconf==2.0.0

--- a/requirements.txt
+++ b/requirements.txt
@@ -10,9 +10,13 @@ certifi==2020.12.5
     # via
     #   -r requirements.in
     #   requests
+cffi==1.14.4
+    # via cryptography
 chardet==3.0.4
     # via requests
-dataclasses==0.7
+cryptography==3.4.7
+    # via -r requirements.in
+dataclasses==0.8
     # via omegaconf
 httplib2==0.19.0
     # via -r requirements.in
@@ -26,6 +30,8 @@ omegaconf==2.0.0
     # via -r requirements.in
 psycopg2-binary==2.8.5
     # via -r requirements.in
+pycparser==2.20
+    # via cffi
 pyparsing==2.4.7
     # via httplib2
 pyyaml==5.4.1

--- a/tests/unit/test_db_config_options.py
+++ b/tests/unit/test_db_config_options.py
@@ -1,10 +1,25 @@
+import datetime
 import os
 import unittest
 
-from buildpack.runtime_components.database import get_config
+from urllib.parse import parse_qs, urlparse, urlencode, urlunparse
+from cryptography.hazmat.primitives import hashes, serialization
+from cryptography.hazmat.primitives.asymmetric import rsa
+from cryptography import x509
+from cryptography.x509 import NameAttribute
+from cryptography.x509.base import Certificate
+from cryptography.x509.oid import NameOID
+from buildpack.runtime_components.database import (
+    get_config,
+    UrlDatabaseConfiguration,
+)
 
 
 class TestDatabaseConfigOptions(unittest.TestCase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.cert_map = {}
+
     def clean_env(self):
 
         # Setting different environment variables for test in the same process
@@ -48,6 +63,41 @@ class TestDatabaseConfigOptions(unittest.TestCase):
         assert config
         assert config["DatabaseType"] == "PostgreSQL"
 
+    def test_inline_certs(self):
+        self.cert_map = CertGen().cert_map
+        self.clean_env()
+        c = UrlDatabaseConfiguration
+        native_params = {
+            c.SSLCERT: self.get_cert("postgresql.crt"),
+            c.SSLROOTCERT: self.get_cert("root.crt"),
+            c.SSLKEY: self.get_cert("postgresql.rsa.key"),
+        }
+        parts = urlparse("postgres://user:secret@host/database")
+        parts = parts._replace(query=urlencode(native_params))
+        native_url = urlunparse(parts)
+        os.environ["DATABASE_URL"] = native_url
+
+        config = get_config()
+        assert config
+        assert config["DatabaseType"] == "PostgreSQL"
+        native_params[c.SSLKEY] = self.get_cert("postgresql.pk8")
+        jdbc_params = parse_qs(urlparse(config["DatabaseJdbcUrl"]).query)
+        self.cmp_cert(native_params, jdbc_params, c.SSLCERT)
+        self.cmp_cert(native_params, jdbc_params, c.SSLROOTCERT)
+        self.cmp_cert(native_params, jdbc_params, c.SSLKEY)
+
+    def get_cert(self, cert_resource):
+        return self.cert_map[cert_resource]
+
+    @classmethod
+    def cmp_cert(cls, native_params, jdbc_params, param):
+        expected_string = native_params[param]
+        actual_file = jdbc_params[param][0]
+        with open(actual_file, "rb") as io_actual:
+            actual_string = io_actual.read().decode("iso-8859-1")
+            assert expected_string == actual_string, param + " differ"
+        os.remove(actual_file)
+
     def test_vcap(self):
         self.clean_env()
         os.environ[
@@ -84,3 +134,97 @@ class TestDatabaseConfigOptions(unittest.TestCase):
         config = get_config()
         assert config
         assert config["DatabaseType"] == "PostgreSQL"
+
+
+# Class to generate a test certificate chain
+# https://cryptography.io/en/latest/x509/tutorial/
+class CertGen:
+    def __init__(self):
+        self.init_root_cert()
+        self.init_postgresql_cert()
+        self.dump_to_storage()
+
+    def dump_to_storage(self):
+        self.cert_map = {}
+        self._dump_cert(self.root_cert, "root.crt")
+        self._dump_cert(self.postgresql_cert, "postgresql.crt")
+        self._dump_key(
+            self.postgresql_key,
+            "postgresql.rsa.key",
+            serialization.Encoding.PEM,
+            serialization.PrivateFormat.TraditionalOpenSSL,
+        )
+        self._dump_key(
+            self.postgresql_key,
+            "postgresql.pk8",
+            serialization.Encoding.DER,
+            serialization.PrivateFormat.PKCS8,
+        )
+
+    def _dump_key(self, key, keyout_name, enc, fmt):
+        self.cert_map[keyout_name] = key.private_bytes(
+            encoding=enc,
+            format=fmt,
+            encryption_algorithm=serialization.NoEncryption(),
+        ).decode("iso-8859-1")
+
+    def _dump_cert(self, cert: Certificate, out_name):
+        self.cert_map[out_name] = cert.public_bytes(
+            serialization.Encoding.PEM
+        ).decode("iso-8859-1")
+
+    def init_root_cert(self):
+        self.root_key = self._newkey()
+        ca_subj = x509.Name(
+            [
+                NameAttribute(NameOID.COUNTRY_NAME, u"US"),
+                NameAttribute(NameOID.ORGANIZATION_NAME, u"Authority, Inc"),
+                NameAttribute(NameOID.COMMON_NAME, u"Authority CA"),
+            ]
+        )
+        self.root_cert = self._sign(
+            ca_subj, self.root_key, ca_subj, self.root_key.public_key(), 3651
+        )
+
+    def init_postgresql_cert(self) -> Certificate:
+        self.postgresql_key = self._newkey()
+        subj = x509.Name(
+            [
+                NameAttribute(NameOID.COUNTRY_NAME, u"US"),
+                NameAttribute(NameOID.ORGANIZATION_NAME, u"Authority, Inc"),
+                NameAttribute(NameOID.COMMON_NAME, u"SQL Client"),
+            ]
+        )
+        self.postgresql_cert = self._sign(
+            self.root_cert.subject,
+            self.root_key,
+            subj,
+            self.postgresql_key.public_key(),
+            3650,
+        )
+
+    @classmethod
+    def _newkey(cls):
+        # Generate our key
+        return rsa.generate_private_key(public_exponent=65537, key_size=2048,)
+
+    @classmethod
+    def _sign(
+        cls, issuer: x509.Name, ca_key, subject: x509.Name, req_pub_key, days
+    ) -> Certificate:
+        # pylint: disable=too-many-arguments
+        return (
+            x509.CertificateBuilder()
+            .subject_name(subject)
+            .issuer_name(issuer)
+            .public_key(req_pub_key)
+            .serial_number(x509.random_serial_number())
+            .not_valid_before(datetime.datetime.utcnow())
+            .not_valid_after(
+                # Our certificate will be valid for 10 days
+                datetime.datetime.utcnow()
+                + datetime.timedelta(days=days)
+                # Sign our certificate with our private key
+            )
+            .sign(ca_key, hashes.SHA256())
+        )

